File: 3225-length-of-longest-subarray-with-at-most-k-frequency/3225-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int res = 0, start = 0, n = nums.length;
        for (int i = 0; i < n; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            while (map.get(nums[i]) > k) {
                map.put(nums[start], map.get(nums[start++]) - 1);
            }
            res = Math.max(res, i - start + 1);
        }
        return res;
    }
}

File: 3225-length-of-longest-subarray-with-at-most-k-frequency/3225-length-of-longest-subarray-with-at-most-k-frequency.kt
class Solution {
    fun maxSubarrayLength(nums: IntArray, k: Int): Int {
        val map = mutableMapOf<Int, Int>()
        var start = 0
        var res = 0
        for (i in nums.indices) {
            map[nums[i]] = map.getOrDefault(nums[i], 0) + 1
            while (map[nums[i]]!! > k) map[nums[start]] = map[nums[start++]]!! - 1
            res = res.coerceAtLeast(i - start + 1)
        }
        return res
    }
    

}